File: Numeros Pseudo-Aleatorios/Tiny Mersenne Twister/tinymt32.scala
/*
 PRNG .......: Tiny Mersenne Twister (TMT)
 Ano ........: 2015 (version 1.1)
 Autor ......: Mutsuo Saito & Makoto Matsumoto
*/

import java.util.concurrent.TimeUnit

class Tinymt32PRNG {
  private var status: Array[Int] = new Array[Int](4)
  private var matrixAlpha: Int = 0
  private var matrixBeta: Int = 0
  private var tempCoeff: Int = 0

  private def tinymt32NextState(): Unit = {
    val stateX: Int = status(0) & 0x7fffffff ^ status(1) ^ status(2)
    val stateY: Int = status(3)
    status(0) = status(1)
    status(1) = status(2)
    status(2) = stateX ^ (stateY << 10)
    status(3) = stateY
    if ((stateY & 1) != 0) {
      status(1) ^= matrixAlpha
      status(2) ^= matrixBeta
    }
  }

  private def tinymt32Temp(): Int = {
    val tempCalc: Int = status(0) + (status(2) >> 8)
    val finalValue: Int = status(3) ^ tempCalc
    finalValue ^ -((tempCalc & 1) & 0xffffffffL).intValue() & tempCoeff
  }

  def tinymt32Init(seed: Int): Unit = {
    status(0) = seed
    matrixAlpha = 0x8f7011ee
    matrixBeta = 0xfc78ff1f
    tempCoeff = 0x3793fdff
    for (i <- 1 to 7) {
      status(i & 3) ^= i + 1812433253 * (status((i - 1) & 3) ^ (status((i - 1) & 3) >>> 30))
    }
    for (_ <- 0 to 7) {
      tinymt32NextState()
    }
  }

  def tinymt32Gen(): Int = {
    tinymt32NextState()
    tinymt32Temp() & 0x7fffffff
  }
}

object Main {
  def main(args: Array[String]): Unit = {
    val seed1: Tinymt32PRNG = new Tinymt32PRNG()
    val seed2: Tinymt32PRNG = new Tinymt32PRNG()

    seed1.tinymt32Init(1234)
    seed2.tinymt32Init((System.currentTimeMillis() / 1000).toInt)

    println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
    println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
    println("PADRAO: Tiny Mersenne Twister (TMT)")
    println("==========================================")
    println()
    println("    Semente: 1234        Semente: TIME")
    println("    -------------        -------------")

    for (_ <- 1 to 10) {
      val Xn1: Int = seed1.tinymt32Gen()
      val Xn2: Int = seed2.tinymt32Gen()
      println(f"${Xn1}%17d${Xn2}%21d")
    }
  }
}

File: Numeros Pseudo-Aleatorios/Tiny Mersenne Twister/tinymt32.java
/*
 PRNG .......: Tiny Mersenne Twister (TMT)
 Ano ........: 2015 (version 1.1)
 Autor ......: Mutsuo Saito & Makoto Matsumoto
*/

import java.util.Random;

class Tinymt32PRNG {
    private int[] status = new int[4];
    private int matrixAlpha;
    private int matrixBeta;
    private int tempCoeff;

    private void tinymt32NextState() {
        int stateX, stateY;
        stateY = status[3];
        stateX = (status[0] & 0x7fffffff) ^ status[1] ^ status[2];
        stateX ^= (stateX << 1);
        stateY ^= (stateY >> 1) ^ stateX;
        status[0] = status[1];
        status[1] = status[2];
        status[2] = stateX ^ (stateY << 10);
        status[3] = stateY;
        if ((stateY & 1) != 0) {
            status[1] ^= matrixAlpha;
            status[2] ^= matrixBeta;
        }
    }

    private int tinymt32Temp() {
        int finalValue, tempCalc;
        finalValue = status[3];
        tempCalc = status[0] + (status[2] >> 8);
        finalValue ^= tempCalc;
        finalValue ^= -((tempCalc & 1) & 0xffffffff) & tempCoeff;
        return finalValue;
    }

    public void tinymt32Init(int seed) {
        status[0] = seed;
        matrixAlpha = 0x8f7011ee;
        matrixBeta = 0xfc78ff1f;
        tempCoeff = 0x3793fdff;
        for (int i = 1; i <= 7; i++) {
            status[i & 3] ^= i + 1812433253 * (status[(i - 1) & 3] ^ (status[(i - 1) & 3] >>> 30));
        }
        for (int i = 0; i < 7; i++) {
            tinymt32NextState();
        }
    }

    public int tinymt32Gen() {
        tinymt32NextState();
        return tinymt32Temp();
    }
}

public class Main {
    public static void main(String[] args) {
        Tinymt32PRNG seed1 = new Tinymt32PRNG();
        Tinymt32PRNG seed2 = new Tinymt32PRNG();

        seed1.tinymt32Init(1234);
        seed2.tinymt32Init((int) (System.currentTimeMillis() / 1000L));

        System.out.println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)");
        System.out.println("PSEUDORANDOM NUMBER GENERATOR (PRNG)");
        System.out.println("PADRAO: Tiny Mersenne Twister (TMT)");
        System.out.println("==========================================");
        System.out.println();
        System.out.println("    Semente: 1234        Semente: TIME");
        System.out.println("    -------------        -------------");

        for (int i = 1; i <= 10; i++) {
            int Xn1 = seed1.tinymt32Gen();
            int Xn2 = seed2.tinymt32Gen();
            System.out.printf("%17d%21d%n", Xn1 & 0xffffffffL, Xn2 & 0xffffffffL);
        }
    }
}
